adapters/repos/db/index.go

```diff
@@ -638,46 +638,54 @@ func (i *Index) IncomingExists(ctx context.Context, shardName string,
 }
 
 func (i *Index) objectSearch(ctx context.Context, limit int,
-	filters *filters.LocalFilter, keywordRanking *traverser.KeywordRankingParams,
+	filters *filters.LocalFilter, keywordRanking *searchparams.KeywordRanking,
 	additional additional.Properties) ([]*storobj.Object, error) {
 	shardNames := i.getSchema.ShardingState(i.Config.ClassName.String()).
 		AllPhysicalShards()
 
-	if len(shardNames) > 1 && keywordRanking != nil {
-		return nil, errors.Errorf("bm25 support limited to single-shard setups for now." +
-			" Multi-shard support expected in v1.13.0")
-	}
-
-	out := make([]*storobj.Object, 0, len(shardNames)*limit)
+	outObjects := make([]*storobj.Object, 0, len(shardNames)*limit)
+	outScores := make([]float32, 0, len(shardNames)*limit)
 	for _, shardName := range shardNames {
 		local := i.getSchema.
 			ShardingState(i.Config.ClassName.String()).
 			IsShardLocal(shardName)
 
-		var res []*storobj.Object
+		var objs []*storobj.Object
+		var scores []float32
 		var err error
 
 		if local {
 			shard := i.Shards[shardName]
-			res, err = shard.objectSearch(ctx, limit, filters, keywordRanking, additional)
+			objs, scores, err = shard.objectSearch(ctx, limit, filters, keywordRanking, additional)
 			if err != nil {
 				return nil, errors.Wrapf(err, "shard %s", shard.ID())
 			}
 
 		} else {
-			res, _, err = i.remote.SearchShard(ctx, shardName, nil, limit, filters, additional)
+			objs, scores, err = i.remote.SearchShard(
+				ctx, shardName, nil, limit, filters, keywordRanking, additional)
 			if err != nil {
 				return nil, errors.Wrapf(err, "remote shard %s", shardName)
 			}
 		}
-		out = append(out, res...)
+		outObjects = append(outObjects, objs...)
+		outScores = append(outScores, scores...)
+	}
+
+	if keywordRanking != nil {
+		res := &inverted.RankedResults{
+			Objects: outObjects,
+			Scores:  outScores,
+		}
+		sort.Sort(res)
+		outObjects = res.Objects
 	}
 
-	if len(out) > limit {
-		out = out[:limit]
+	if len(outObjects) > limit {
```

Comment:
is here a possibility that passed limit can be negative?

Reply:
If a negative limit was passed, this function, [db.getLimit](https://github.com/semi-technologies/weaviate/blob/26cd4f787861dd88dd688c7d716943f10a767b26/adapters/repos/db/search.go#L260) will have already set the limit to the default by the time the thread of execution reaches this point

Reply:
but doesn't this method gets the limit by invoking [db.getTotalLimit](https://github.com/semi-technologies/weaviate/blob/26cd4f787861dd88dd688c7d716943f10a767b26/adapters/repos/db/search.go#L214) ?

Reply:
`db.getTotalLimit` just wraps `db.getLimit`. So that:

- if a negative limit was passed to Get query
- if there are no "near" params

`db.getLimit`, and therefore `db.getTotalLimit` will return a non-zero value